src/ApplicationStatus.cpp

```diff
@@ -0,0 +1,210 @@
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if (m_shmem->size() != buffer_size(m_num_cpu)) {
+            throw Exception("ApplicationStatus: shared memory incorrectly sized",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        // Note: no lock; all members of the struct are 32-bits and will be
+        // accessed atomically by hardware.
+        m_buffer = (m_app_status_s *)m_shmem->pointer();
+
+        // initialize shmem if all zero is not appropriate
+        for (int ii = 0; ii < m_num_cpu; ++ii) {
+            m_buffer[ii].total_work = -1;
+            m_buffer[ii].process = -1;
+        }
+    }
+
+    void ApplicationStatusImp::set_hint(int cpu_idx, uint64_t hints)
+    {
+        if (cpu_idx < 0 || cpu_idx >= m_num_cpu) {
+            throw Exception("ApplicationStatusImp::set_hint(): invalid CPU index: " + std::to_string(cpu_idx),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        if ((hints & ~GEOPM_MASK_REGION_HINT) != 0ULL) {
+            throw Exception("ApplicationStatusImp::set_hint(): invalid hint",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        GEOPM_DEBUG_ASSERT(m_buffer != nullptr, "m_buffer not set");
+        // pack hints into 32 bits for atomic write
+        m_buffer[cpu_idx].hints = (uint32_t)(hints >> 32);
+    }
+
+    uint64_t ApplicationStatusImp::get_hint(int cpu_idx) const
+    {
+        if (cpu_idx < 0 || cpu_idx >= m_num_cpu) {
+            throw Exception("ApplicationStatusImp::get_hint(): invalid CPU index: " + std::to_string(cpu_idx),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        GEOPM_DEBUG_ASSERT(m_buffer != nullptr, "m_buffer not set");
+        return (uint64_t)m_buffer[cpu_idx].hints << 32;
+    }
+
+    std::vector<uint64_t> ApplicationStatusImp::get_hint(void) const
+    {
+        GEOPM_DEBUG_ASSERT(m_buffer != nullptr, "m_buffer not set");
+        std::vector<uint64_t> result(m_num_cpu);
+        for (int ii = 0; ii < m_num_cpu; ++ii) {
+            result[ii] = get_hint(ii);
+        }
+        return result;
+    }
+
+    void ApplicationStatusImp::set_total_work_units(int cpu_idx, int work_units)
+    {
+        if (cpu_idx < 0 || cpu_idx >= m_num_cpu) {
+            throw Exception("ApplicationStatusImp::set_total_work_units(): invalid CPU index: " + std::to_string(cpu_idx),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        GEOPM_DEBUG_ASSERT(m_buffer != nullptr, "m_buffer not set");
+
+        if (work_units < 0) {
+            throw Exception("ApplicationStatusImp::set_total_work_units(): invalid number of work units: " + std::to_string(work_units),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+
+        }
+
+        m_buffer[cpu_idx].total_work = work_units;
+        m_buffer[cpu_idx].completed_work = 0;
```

Comment:
set completed_work to zero prior to setting total_work units to avoid reporting more complete than total.